File: java/medium/BearAndSteadyGene.java
// https://www.hackerrank.com/challenges/bear-and-steady-gene/problem

package java.medium;

import java.util.*;

class BearAndSteadyJene {

    public static int steadyGene(String gene){

        int n=gene.length();

        HashMap<Character,Integer> map=
                new HashMap<>();

        for(char c:gene.toCharArray())
            map.put(c,
            map.getOrDefault(c,0)+1);

        int req=n/4;

        int left=0;

        int ans=n;

        for(int right=0;right<n;right++){

            char ch=gene.charAt(right);

            map.put(
            ch,
            map.get(ch)-1);

            while(
                left<n &&
                map.getOrDefault('A',0)<=req &&
                map.getOrDefault('C',0)<=req &&
                map.getOrDefault('G',0)<=req &&
                map.getOrDefault('T',0)<=req
            ){

                ans=Math.min(
                ans,
                right-left+1);

                char c=
                gene.charAt(left);

                map.put(
                c,
                map.get(c)+1);

                left++;
            }
        }

        return ans;
    }
}

File: js/medium/bear-and-steady-gene.js
function steadyGene(gene){

    let n=gene.length;

    let cnt=new Map();

    for(let c of gene)
        cnt.set(
        c,
        (cnt.get(c)||0)+1
        );

    let req=n/4;

    let left=0;

    let ans=n;

    for(let right=0;
        right<n;
        right++){

        let ch=gene[right];

        cnt.set(
        ch,
        cnt.get(ch)-1
        );

        while(
        (cnt.get('A')||0)<=req &&
        (cnt.get('C')||0)<=req &&
        (cnt.get('G')||0)<=req &&
        (cnt.get('T')||0)<=req
        ){

            ans=Math.min(
            ans,
            right-left+1
            );

            cnt.set(
            gene[left],
            cnt.get(
            gene[left]
            )+1
            );

            left++;
        }
    }

    return ans;
}